### PR TITLE
i#5994 opcode_mix, part1: add categories to opcode_mix output

### PR DESCRIPTION
**Pull request by edeiana**
Groups instructions that belong to the same categories together, and prints (to std::cerr) a count of them at the end of opcode counts.

Issue: #5994

### CODE REVIEW

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.cpp:236]: Decoding is expensive and we took some effort above to cache the results, so let's leverage the same caching by using the same decode call as above and storing the categories in the (maybe renamed) opcode_cache.

File: core/ir/instr_api.h
Comment (derekbruening) [inline core/ir/instr_api.h:743]: Hmm, it seems more flexible to instead provide a string name for an individual category so the user can combine them in whatever way they want, rather than hardcoding commas or a certain ordering?

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:109]: Please add a test.  Maybe something like:
```
      torunonly_drcacheoff(opcode_categories allasm_x86_64
        "" "@-simulator_type@opcode_mix" "")
```
That tiny app has a small fixed instr sequence so you can hardcode the category output into a template file `clients/drcachesim/tests/offline-opcode_categories.templatex`.

**Review by derekbruening — COMMENTED** [review]
Thank you for contributing!  I have some suggestions.

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.cpp]: Done.

File: clients/drcachesim/tools/opcode_mix.h
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.h:110]: Done.

File: clients/drcachesim/tests/offline-opcode_categories.templatex
Comment (derekbruening) [inline clients/drcachesim/tests/offline-opcode_categories.templatex:30]: (Interesting to see category results: this is going to be a good tool addition.

Hmm, to get 33 it must count `syscall` in addition to `jnz`?  I guess a system call gateway is a kind of branch.  Would everyone call it a branch?  Sorry, I realize this discussion is not really in the scope of this PR: just interesting; not clearly a problem so I guess we leave it and see if someone complains.)

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:87]: Hmm I was going to say `s/uint/dr_instr_category_t/` but this combines them.  Add a comment on why it's a uint and not the enum type?

File: core/ir/instr_api.h
Comment (derekbruening) [inline core/ir/instr_api.h:2551]: (I guess your clang-format is different -- but if the GA CI passed someone else would hit this anyway as it must be an older version vs newer.  Small enough to not need separation/tweaking; seems ok to leave in this PR.)

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:87]: (Optional (wish Github had a way to mark things as optional): use a new struct with meaningful names rather than std::pair.  This is not in our Github style guide so optional.)

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.cpp:273]: How about a loop that marches through all the bits, to generalize this code and avoid us having to edit it when a new category is added?   Maybe don't even need a max named bit: going to the 32nd bit seems ok to future-proof.

**Review by derekbruening — COMMENTED** [review]
Looking good, but I had a few more suggestions.

File: core/ir/instr_shared.c
Comment (derekbruening) [inline core/ir/instr_shared.c:474]: We should have at least a sanity test of a new API in the DR core's tests (clients/drcachesim is considered separate from the "core").  Please add some spot-checks of category strings, maybe in test_categories() in suite/tests/api/drdecode_x86.c.

Comment (derekbruening) [inline core/ir/instr_shared.c:474]: Please add the new API to the changelist at https://github.com/DynamoRIO/dynamorio/blob/master/api/docs/release.dox#L195

**Review by derekbruening — COMMENTED** [review]
Two more requests.  (Sorry, not trying to "move the goalposts": should have put these in the first round of reviews.)

File: clients/drcachesim/tests/offline-opcode_categories.templatex
Comment (edeiana) [reply clients/drcachesim/tests/offline-opcode_categories.templatex:30]: Right, https://github.com/DynamoRIO/dynamorio/blob/master/core/ir/x86/decode_table.c#L2467 categorized `syscall` as a branch.
Not really sure if that's a good semantics, but as pointed out, it's outside the scope of this PR.

File: clients/drcachesim/tools/opcode_mix.h
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.h]: Done, I changed it to a more meaningful structure as suggested.

File: core/ir/instr_api.h
Comment (edeiana) [reply core/ir/instr_api.h:2551]: Yes, this is a clang-format issue. I'm using version 9.0.0, not sure if that's old or new in tis context. I can always revert back if you'd like me to!

File: clients/drcachesim/tools/opcode_mix.h
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.h]: Done.

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.cpp]: Whoops, this makes so much more sense. Done.

File: core/ir/instr_shared.c
Comment (edeiana) [reply core/ir/instr_shared.c:474]: Done. I added the check in the CATEGORY_CHECK macro, but if you want me to separate the two and add more "spot-checks" for more instructions I can do that.

Comment (edeiana) [reply core/ir/instr_shared.c:474]: Done.

File: core/ir/instr_api.h
Comment (derekbruening) [reply core/ir/instr_api.h:2551]: We have documented that we're using version 14: https://dynamorio.org/page_code_style.html#autotoc_md122
Given that yours is older, yes please revert.  (If yours was newer we might consider updating.)

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.cpp:231]: This line looks too long: shouldn't clang-format have split it at 90?

File: core/ir/instr_shared.c
Comment (derekbruening) [reply core/ir/instr_shared.c:474]: Maybe add one instance of a multi-category instruction?

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.cpp]: Yes, made me realize I had broken my vim config and clang-format wasn't running after saving. Fixed, thanks!

File: core/ir/instr_api.h
Comment (edeiana) [reply core/ir/instr_api.h]: Sorry, missed that. Fixed.

File: core/ir/instr_shared.c
Comment (edeiana) [reply core/ir/instr_shared.c:474]: Done. Added one for "uncategorized" as well, since it's kind of an edge case being all 0s.

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [inline suite/tests/api/drdecode_x86.c:45]: stderr seems better for errors.

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [inline suite/tests/api/drdecode_x86.c:274]: Could have remained as it was?  (Aside for future PR's on other tests: Note that tests that interleave output with DR's messages to stderr would want stderr for everything, and they use "print()" from tools.h but this standalone app test does not.)

**Review by derekbruening — APPROVED** [review]

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c:45]: I agree, I just wanted to preserve the same behavior to avoid "breaking" anything.
The reason I changed it to `fprintf` is that I included `suite/tests/tools.h` to get access to `TESTANY` in `CHECK_CATEGORY`, but `tools.h` has https://github.com/DynamoRIO/dynamorio/blob/e4da1a7711c1f1dc1bb5757445d3e1097837177d/suite/tests/tools.h#L360 which prevents using `printf`.
I'll change it to stderr (similar to other `ASSERT` I've seen: https://github.com/DynamoRIO/dynamorio/blob/e4da1a7711c1f1dc1bb5757445d3e1097837177d/suite/tests/api/ir_x86.c#L71)

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c:274]: Same reason of https://github.com/DynamoRIO/dynamorio/pull/6512#discussion_r1463943359 .
Should I change it to `print` now that `tools.h` is included? (Or was I not supposed to include `tools.h` here?)

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [reply suite/tests/api/drdecode_x86.c:45]: Oh if tools.h is successfully included then use print()

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [reply suite/tests/api/drdecode_x86.c:274]: If tools.h is successfully included then use print().  There is some build config complexity with tools.h vs config.h vs dr_api.h but if it worked then we're good.

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c]: Done. Using `print()` also in this `ASSERT`.

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c]: Done. The test is built correctly, and the print statements work, at least when I run the `code_api|api.drdecode` test.

**Review by derekbruening — APPROVED** [review]

**Comment by edeiana** [comment]
Test opcode_categories fixed.
It was in the wrong place in suite/tests/CMakeLists.txt.
However, a few (minor) things changed since last approval, so I'd like a second pair of eyes before merging into master.

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.cpp:272]: This could be a constant (either `const` or `#define`, and name in all-caps).

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:88]: s/0/OP_INVALID/

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:89]: s/0/DR_INSTR_CATEGORY_UNCATEGORIZED/

File: clients/drcachesim/tools/opcode_mix.h
Comment (derekbruening) [inline clients/drcachesim/tools/opcode_mix.h:108]: Optional: "opcode_data_cache" or "opcode_info_cache" sounds more like it includes several pieces of info and not just a category.  On that same note: "opcode_category_data_t" sounds like just the category and maybe "opcode_data_t" or "opcode_info_t" is more descriptive?  It's subjective though: if you like it how it is better please keep it.

File: core/ir/instr_shared.c
Comment (derekbruening) [inline core/ir/instr_shared.c:477]: Add a (non-Doxygen) comment by the dr_instr_category_t enum as a reminder to update this function when adding new categories.

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [inline suite/tests/api/drdecode_x86.c:161]: Putting key steps needed for subsequent code inside something like ASSERT which is sometimes debug-only and turns into a nop in release build can be dangerous as the subsequent code will then fail.  Here it looks like ASSERT is defined in release build too.  But imagine this code being copied/modeled in some other file where that's not the case.  So it may be worth making safer code by separating the instr_encode and decode calls from the asserts.  Or add a comment stating the assumption that ASSERT is always enabled.

File: suite/tests/api/drdecode_x86.c
Comment (derekbruening) [inline suite/tests/api/drdecode_x86.c:185]: Can be `const`.  Ditto below.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tools/opcode_mix.cpp
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.cpp:272]: Done.

File: clients/drcachesim/tools/opcode_mix.h
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.h:88]: Done.

File: clients/drcachesim/tools/opcode_mix.h
Comment (edeiana) [reply clients/drcachesim/tools/opcode_mix.h:108]: I like the "data" variant. Done.

File: core/ir/instr_shared.c
Comment (edeiana) [reply core/ir/instr_shared.c:477]: Done.

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c:161]: Done.

File: suite/tests/api/drdecode_x86.c
Comment (edeiana) [reply suite/tests/api/drdecode_x86.c:185]: Done.

**Pull request merged**
